grpc/src/main/java/io/grpc/netty/shaded/io/grpc/netty/CustomChannelFactory.java

```diff
@@ -0,0 +1,64 @@
+package io.grpc.netty.shaded.io.grpc.netty;
+
+import io.grpc.netty.shaded.io.netty.channel.Channel;
+import io.grpc.netty.shaded.io.netty.channel.ChannelFactory;
+import io.grpc.netty.shaded.io.netty.channel.ServerChannel;
+import io.grpc.netty.shaded.io.netty.channel.group.ChannelGroup;
+import io.grpc.netty.shaded.io.netty.channel.group.DefaultChannelGroup;
+import io.grpc.netty.shaded.io.netty.util.concurrent.GlobalEventExecutor;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Predicate;
+
+/**
+ * Tracks all the created sessions. It wraps the {@link Utils#DEFAULT_SERVER_CHANNEL_FACTORY} and
+ * uses it for session creation. Once it's notified via the {@code closeFilter()} method, it closes
+ * all channels matching the predicate.
+ */
+public class CustomChannelFactory implements ChannelFactory<ServerChannel> {
+
+  private final ChannelFactory<? extends ServerChannel> channelFactory;
+  private final List<Channel> channels = new ArrayList<>();
+
+  public CustomChannelFactory() {
+    this.channelFactory = Utils.DEFAULT_SERVER_CHANNEL_FACTORY;
+  }
+
+  @Override
+  public ServerChannel newChannel() {
+    ServerChannel serverChannel = channelFactory.newChannel();
+    channels.add(serverChannel);
+    return serverChannel;
+  }
+
+  public void closeFilter(Predicate<Map<String, String>> headerFilter) {
+    ChannelGroup allChannels = new DefaultChannelGroup(GlobalEventExecutor.INSTANCE);
+    channels.stream()
+        .filter(
+            channel -> {
+              ProxyInfo proxyInfo = channel.attr(ProxyInfo.attributeKey).get();
```

Comment:
Where's the proxy info set? As far as I can tell it's not being set, but maybe I'm missing something. This needs to be with the correct values for the filter to correctly identify the connections to close. 

Also, maybe we use the gRPC metadata (header) values here instead of using `ProxyInfo`.

Reply:
You were right, I added logs and getting:
```
ava:59 - proxyInfo:null
INFO  [topology-tracker-listener-executor@024d599d-192e-42dd-8e88-3c653f33d869] 2021-12-01 13:01:10,374 CustomChannelFactory.java:62 - headers: {}
INFO  [topology-tracker-listener-executor@024d599d-192e-42dd-8e88-3c653f33d869] 2021-12-01 13:01:10,375 CustomChannelFactory.java:74 - Closing channels, number of channels: 0 
```
However, what is weird, the code from this PR does not have the problem during unassign. 
Maybe it is related to increasing `<grpc.version>1.42.1</grpc.version>`. I need to test this change in isolation.

Reply:
Good news, when testing this: https://github.com/stargate/stargate/pull/1447/files
in isolation, the problem with unassign is not appearing